# <center> TP N°1 </center>
### <center> Methodes Directes pour la Résolution des Systèmes Linéaires Ax = b </center>
<div>
    <center> RAZAFINDRAZAKA Henintsoa </center>
    <center> Wang James </center>
    <center> ____________</center>
    <center> ING5 SI GR02 </center>
    <center> 19/10/2020 </center>
</div>

1. Ouvrir un fichier qu’on nomme `trianinf.m`. Dans ce fichier, on va créer la fonction Matlab qui à partir d’une matrice inférieure T et d’un second membre b, résout le système Tx = b. La syntaxe de début du programme est la suivante :

```
function x=trianginf(T,b)
% T est la matrice triangulaire inférieure et b le vecteur 2nd membre.
n=length(b);
x(1)=
for i=2:n
x(i)=...
end
```

In [1]:
import numpy as np

n = 4
T = np.tril(np.random.randint(10,size=(n,n)) )

while np.linalg.det(T) == 0.0: 
    T = np.tril(np.random.randint(10,size=(n,n)) )

print(T)
print('\n|T|= ', np.linalg.det(T))

[[8 0 0 0]
 [8 6 0 0]
 [9 4 7 0]
 [0 4 2 3]]

|T|=  1007.9999999999993


In [2]:
b = np.random.randint(10, size=(n,))
b

array([7, 8, 9, 7])

In [3]:
# Debut de la fonction trianginf(T,b):
def trianginf(T,b):
    n = len(b)
    x = np.zeros(b.shape)

    x[0] = b[0] / T[0][0]
    for i in range(1,n):
        sigma = 0
        for j in range(0, i):
            sigma += T[i][j] * x[j]
        x[i] = (1/T[i][i]) * ( b[i] - sigma )
    return x

x_test = trianginf(T,b)
T.dot(x_test)

array([7., 8., 9., 7.])

2. Ouvrir un fichier qu’on nomme `triansup.m`. Dans ce fichier, on va créer la fonction Matlab qui à partir d’une matrice supérieure T et d’un second membre b, résout le système T x = b. La syntaxe de début du programme est la suivante :

```
function x=triangsup(T,b)
% T est la matrice triangulaire supérieure et b le vecteur 2nd membre.
n=length(b);
x(n)=
for i=n-1:-1:1 % boucle incrément négatif
x(i)=...
end
```

In [4]:
n = 4
T = np.triu(np.random.randint(10,size=(n,n)) )

while np.linalg.det(T) == 0.0: 
    T = np.tril(np.random.randint(10,size=(n,n)) )

print(T)
print('\n|T|= ', np.linalg.det(T))

[[4 9 6 3]
 [0 9 8 8]
 [0 0 9 1]
 [0 0 0 6]]

|T|=  1943.9999999999984


In [5]:
b = np.random.randint(10, size=(n,))
b

array([0, 8, 7, 6])

In [6]:
# Debut de la fonction triansup(T,b):
def triangsup(T,b):
    n = len(b)
    x = np.zeros(b.shape)

    x[n-1] = b[n-1] / T[n-1][n-1]
    for i in range(n-2,-1,-1):
        sigma = 0
        for j in range(i+1, n):
            sigma += T[i][j] * x[j]
        x[i] = (1/T[i][i]) * ( b[i] - sigma )
    return x

x = triangsup(T,b)
print(np.linalg.cond(T))
print(b)
print(T.dot(x))

8.635292274040966
[0 8 7 6]
[0. 8. 7. 6.]


3. Ouvrir un fichier qu’on nomme `gauss.m`. Dans ce fichier, on va créer la fonction Matlab qui à
partir d’une matrice A effectue de Gauss et résout le système Ax = b. La syntaxe de début du programme est la suivante :

```
function x=gauss(A,b)
% A est la matrice principale et b le vecteur second membre.
n=length(b);
.....
```

In [7]:
A = np.random.rand(n,n)
A_init = A
print(A)
print('\n|T|= ', np.linalg.det(A))


[[0.90423601 0.13350676 0.57021057 0.83274163]
 [0.56346218 0.4737454  0.4027603  0.35088916]
 [0.17974789 0.42524247 0.16434862 0.8644665 ]
 [0.96594981 0.77196819 0.65732671 0.90683765]]

|T|=  0.008945179786478592


In [8]:
b_init = np.random.randint(10, size=(n,))
b_init

array([3, 8, 5, 7])

In [9]:
# triangulation de A
A_init = A
b = b_init
def gauss(A, b):
    m = np.zeros(A.shape)
    for k in range(0, n-1):
        for i in range(k+1, n):
            m[i][k] = A[i][k] / A[k][k]
            for j in range(k, n):
                A[i][j] = A[i][j] - m[i][k]*A[k][j]
            b[i] = b[i] - m[i][k] * b[k]
    
    x = triangsup(A,b)
    return x

x = gauss(A,b)

print("\nA_init*x = ",A_init.dot(x))
print("\nb_init = ",b_init)



A_init*x =  [  3.   6.  -2. -27.]

b_init =  [  3   6  -2 -27]


4. Tester la méthode sur un exemple et comparer votre résultat avec celui obtenu avec la commande A/b et celui obtenu avec la commande inv(A)*b. Que se passe-t-il quand on augmente la taille de la matrice (n = 10, 100, 1000, ...).

In [10]:
import time

myround = np.vectorize(round)

n = 10
while n <= 1000 :

    A_init = np.random.rand(n,n)
    A = A_init

    b_init = np.random.randint(10, size=(n,))
    b = b_init
    
    print('\nn = ', n)
    
    start_time = time.time()
    
    # Notre gauss
    print('\n===========My gauss ============\n', myround(gauss(A_init,b_init)))
    
    print("--- %s seconds ---" % (time.time() - start_time))
    
    start_time = time.time()
    
    # Calcul python A^(-1).b
    A_inv = np.linalg.inv(A_init)
    print('\n=========== A^(-1).b ===========\n',myround( A_inv.dot(b_init) ))
    
    print("--- %s seconds ---" % (time.time() - start_time))
    
    n *= 10


n =  10

===========My gauss ============
 [-14.  16.  -2.  12.  -8.  -5.   2.   7. -15.  24.]
--- 0.002000570297241211 seconds ---

=========== A^(-1).b ===========
 [-14.  16.  -2.  12.  -8.  -5.   2.   7. -15.  24.]
--- 0.005059719085693359 seconds ---

n =  100

===========My gauss ============
 [ 5.  1.  3. -4. -1.  1. -4.  1.  1.  2.  7. -7. -1.  1. -1.  4. -6. -1.
  4. -3.  0. -3.  3.  1. -3.  0. -2.  1. -2.  0.  2.  2.  1.  1. -1. -1.
  0.  4.  2.  2. -1. -1.  2. -2.  2.  2.  1.  1.  1.  1.  0. -2.  0.  1.
  2. -2.  1.  2.  4. -1.  0. -1. -3. -2. -1.  1. -5. -1.  1. -2.  1.  2.
 -1.  0.  0.  0.  0. -1.  1.  1.  1.  0. -1.  0.  0. -1.  0.  0.  0.  0.
 -1.  0.  0.  0.  0.  0.  0.  1.  1. -1.]
--- 0.5241184234619141 seconds ---

=========== A^(-1).b ===========
 [ 5.  1.  3. -4. -1.  1. -4.  1.  1.  2.  7. -7. -1.  1. -1.  4. -6. -1.
  4. -3.  0. -3.  3.  1. -3.  0. -2.  1. -2.  0.  2.  2.  1.  1. -1. -1.
  0.  4.  2.  2. -1. -1.  2. -2.  2.  2.  1.  1.  1.  1.  0. -2.  0.  1.
  

n =  10 
- My gauss : 0.0020008087158203125 secondes 
- A^(-1).b : 0.0 secondes

n =  100
- My gauss : 0.5151150226593018 secondes
- A^(-1).b : 0.002000093460083008 secondes

n =  1000
- My gauss : 479.6726932525635 secondes
- A^(-1).b : 0.05336451530456543 secondes

Lorsqu'on augmente la taille de la matrice n, le temps d'éxecution de la méthode de Gauss augmente exponentiellement contrairement à la méthode inverse. 

5. Ouvrir un fichier qu’on nomme `ludecomp.m`. Dans ce fichier, on va créer la fonction Matlab qui à partir d’une matrice A effectue la factorisation de type LU sous la forme A = LU. La syntaxe de début du programme est la suivante :

```
function [L,U]=ludecomp(A)
% A est la matrice principale.
n=length(b);
.....
```

In [11]:
n = 4
A = np.random.randint(10,size=(n,n))
A_init = A
print(A)
print('\n|T|= ', np.linalg.det(A))

b_init = np.random.randint(10, size=(n,))
print(b_init)

[[5 8 8 4]
 [7 2 0 8]
 [3 8 9 0]
 [8 1 4 1]]

|T|=  -281.99999999999955
[5 8 3 9]


In [19]:
b = b_init

def ludecomp(A):
    n = len(b)
    L = np.zeros(A.shape)
    U = np.zeros(A.shape)
    L[0][0] = 1
    
    for j in range(0, n):
        U[0][j] = A[0][j]
        
    for i in range(1, n):
        L[i][0] = A[i][0] / U[0][0]
        
    for i in range(1, n):
        L[i][i] = 1
        s0 = 0
        for k in range(0, i):
            s0 += L[i][k] * U[k][i]
        U[i][i] = A[i][i] - s0
        
        for j in range(i, n):
            s1 = 0
            for k in range(0, i):
                s1 += L[i][k] * U[k][j]
            U[i][j] = A[i][j] - s1
            
            s2 = 0
            for k in range(0, i):
                s2 += L[j][k] * U[k][i]
            L[j][i] = (1 / U[i][i]) * (A[j][i] - s2)
    return L, U

In [20]:
L, U  = ludecomp(A)

print("L : \n ",L)
print("\n\nU : \n",U)

L : 
  [[ 1.          0.          0.          0.        ]
 [ 1.4         1.          0.          0.        ]
 [ 0.6        -0.34782609  1.          0.        ]
 [ 1.6         1.2826087  18.28571429  1.        ]]


U : 
 [[  5.           8.           8.           4.        ]
 [  0.          -9.2        -11.2          2.4       ]
 [  0.           0.           0.30434783  -1.56521739]
 [  0.           0.           0.          20.14285714]]


6. Ouvrir un fichier qu’on nomme lusovle.m. Dans ce fichier, on va créer la fonction Matlab qui à partir d’une matrice A et d’un vecteur b résout le système Ax = b à partir de la factorisation LU de la matrice A. La syntaxe de début du programme est la suivante:

```
function [L,U]=ludecomp(A)
% A est la matrice principale.
n=length(b);
.....
```

Ax = B ==> LUx = B ==> Ly = B avec Ux = y

In [22]:
y = trianginf(L,b) 
print("y : \n ",y)


x = triangsup(U,y)
print("\nx: \n", x)
print("\nnp.linalg.inv(A_init).dot(b)) : ", np.linalg.inv(A).dot(b))

print("\nb : \n", b)
print("\nA.dot(x) : ", A.dot(x))

y : 
  [ 5.          1.          0.34782609 -6.64285714]

x: 
 [ 1.38297872  0.4787234  -0.55319149 -0.32978723]

np.linalg.inv(A_init).dot(b)) :  [ 1.38297872  0.4787234  -0.55319149 -0.32978723]

b : 
 [5 8 3 9]

A.dot(x) :  [5. 8. 3. 9.]


7. Tester la méthode sur un exemple et comparer votre résultat avec celui obtenu avec la fonction lu déja implémenter dans Matlab (help lu), la commande A/b et celui obtenu avec la commande inv(A)*b. Que se passe-t-il quand on augmente la taille de la matrice (n = 10, 100, 1000, ...).


In [25]:
import time

myround = np.vectorize(round)

n = 10
while n <= 1000 :

    A_init = np.random.rand(n,n)
    A = A_init

    b_init = np.random.randint(10, size=(n,))
    b = b_init
    
    print('\nn = ', n)
    
    start_time = time.time()
    
    # Notre gauss
    print('\n===========My gauss ============\n', myround(gauss(A_init,b_init)))
    
    print("--- %s seconds ---" % (time.time() - start_time))
    
    start_time = time.time()
    
    # Calcul python A^(-1).b
    A_inv = np.linalg.inv(A_init)
    print('\n=========== A^(-1).b ===========\n',myround( A_inv.dot(b_init) ))
    
    print("--- %s seconds ---" % (time.time() - start_time))
    
    start_time = time.time()
    
    # méthode LU
    L, U  = ludecomp(A)
    A_inv = np.linalg.inv(A_init)
    y = trianginf(L,b) 
    x = triangsup(U,y)
    print('\n=========== LU ===========\n',myround(x) )
    
    print("--- %s seconds ---" % (time.time() - start_time))
    
    n *= 10


n =  10

===========My gauss ============
 [ 2.  6. -3.  2. -1. -1.  1.  0.  1.  1.]
--- 0.0010006427764892578 seconds ---

=========== A^(-1).b ===========
 [ 2.  6. -3.  2. -1. -1.  1.  0.  1.  1.]
--- 0.0010001659393310547 seconds ---

=========== LU ===========
 [ 2.  6. -3.  2. -1. -1.  1.  0.  1.  1.]
--- 0.0010004043579101562 seconds ---

n =  100

===========My gauss ============
 [ 5.  2.  2. -1.  3. -6.  3. -5.  3.  4.  1. -4.  3.  1. -2.  4.  0. -2.
  3.  0. -3.  1. -3. -5.  1. -2.  2.  0.  2. -1.  1. -1. -2.  2. -2.  1.
  3.  0. -1. -3.  1.  1.  0.  2.  0.  2.  2. -1.  2.  0.  2.  0.  0.  0.
 -1. -1.  0.  1. -1. -1.  1. -1. -1.  0. -1.  0. -1.  0. -1. -1.  0.  0.
  0.  0. -1.  1.  0.  0. -1.  0.  0.  1.  1.  0.  0.  1.  1.  0.  0.  0.
  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.]
--- 0.5211272239685059 seconds ---

=========== A^(-1).b ===========
 [ 5.  2.  2. -1.  3. -6.  3. -5.  3.  4.  1. -4.  3.  1. -2.  4.  0. -2.
  3.  0. -3.  1. -3. -5.  1. -2.  2.  0.  2. -1.  1. -1. 


=========== LU ===========
 [ 5.  6. -5.  7. -1. -7. -4.  8.  0.  2.  1.  2.  5. -1. -2. -7.  1. -5.
 -4.  0.  0. -1. -2. -3.  2.  1. -1.  1.  3. -6.  7.  3.  4.  3. -3.  0.
 -3. -3. -5.  0.  5. -1.  0.  1.  0.  5.  2.  1. -8. -2.  0.  0. -3.  5.
  6. -4.  5. -2.  2.  0.  3.  3. -1.  1. -4.  1. -6. -4.  5. -1.  2. -5.
  4. -2.  5.  0. -5.  3.  0. -2. -4. -1. -6.  1. -1.  3. -4.  0.  1.  0.
 -1. -3. -2. -4. -2.  3.  2. -4.  2. -4.  2.  1.  1. -5.  2. -1.  1. -2.
  4. -3.  0.  4.  3. -2. -3. -1.  3. -1.  1.  0.  5.  1.  0. -3.  1.  0.
 -1.  0. -1.  1. -3. -2.  2.  0.  1.  0.  2. -2. -2.  2.  0. -1.  2. -3.
  5.  0.  0. -1.  4. -2.  1.  4.  3.  2.  1. -2. -3. -3.  0. -2.  4. -1.
  0. -3.  1. -1.  0.  0. -2.  3. -2.  1. -1.  2. -1. -3.  1. -3. -2. -3.
  2. -1.  3. -4.  1.  1. -2.  1.  0.  1.  2. -3.  4.  0. -1. -3. -2.  0.
  1.  1.  0.  1.  0.  4.  0.  2. -3. -1. -3. -1.  0.  2. -1.  0.  1.  0.
  2.  2.  0. -3.  1.  0.  0.  3.  4.  1.  2.  0. -1. -1.  4. -1. -3.  1.
 -2. -1. -1.  3.  1.  

n =  10 
- My gauss : 0.0010006427764892578 secondes 
- A^(-1).b : 0.0010001659393310547 secondes
- LU : 0.0010004043579101562 secondes

n =  100
- My gauss : 0.5211272239685059  secondes
- A^(-1).b : 0.0020003318786621094 secondes
- LU : 0.35507893562316895 secondes

n =  1000
- My gauss : 487.00597405433655 secondes
- A^(-1).b : 0.048001766204833984 secondes
- LU : 294.45130681991577 secondes

Lorsqu'on augmente la taille de la matrice n, les temps d'éxecution sur les méthodes de décomposition LU et 
de Gauss augmentent exponentiellement contrairement à la méthode inverse. Le temps d'éxecution de LU reste tout de meme inférieur que celui de Gauss. 